airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/csv_parser.py

```diff
@@ -129,96 +174,147 @@ def _to_nullable(row: Mapping[str, str], null_values: Set[str]) -> Dict[str, Opt
+                        warnings.append(_format_warning(key, value, prop_type))
+
+                elif python_type == bool:
+                    try:
+                        cast_value = _value_to_bool(value, config_format.true_values, config_format.false_values)
+                    except ValueError:
+                        warnings.append(_format_warning(key, value, prop_type))
+
+                elif python_type == dict:
+                    try:
+                        # we don't re-use _value_to_object here because we type the column as object as long as there is only one object
+                        cast_value = json.loads(value)
+                    except json.JSONDecodeError:
+                        warnings.append(_format_warning(key, value, prop_type))
+
+                elif python_type == list:
+                    try:
+                        cast_value = _value_to_list(value)
+                    except (ValueError, json.JSONDecodeError):
+                        warnings.append(_format_warning(key, value, prop_type))
+
+                elif python_type:
+                    try:
+                        cast_value = _value_to_python_type(value, python_type)
+                    except ValueError:
+                        warnings.append(_format_warning(key, value, prop_type))
 
+            else:
+                warnings.append(_format_warning(key, value, prop_type))
+
+            result[key] = cast_value
+
+        if warnings:
+            logger.warning(
+                f"{FileBasedSourceError.ERROR_CASTING_VALUE.value}: {','.join([w for w in warnings])}",
+            )
+        return result
+
+
+class _TypeInferrer:
+    _BOOLEAN_TYPE = "boolean"
+    _INTEGER_TYPE = "integer"
+    _NUMBER_TYPE = "number"
+    _ARRAY_TYPE = "array"
+    _OBJECT_TYPE = "object"
+    _STRING_TYPE = "string"
+
+    def __init__(self, boolean_trues: Set[str], boolean_falses: Set[str]) -> None:
+        self._boolean_trues = boolean_trues
+        self._boolean_falses = boolean_falses
+        self._values: Set[str] = set()
+
+    def add_value(self, value: Any) -> None:
+        self._values.add(value)
+
+    def infer(self) -> str:
+        types = {self._infer_type(value) for value in self._values}
+
+        if types == {self._BOOLEAN_TYPE}:
+            return self._BOOLEAN_TYPE
+        elif types == {self._INTEGER_TYPE}:
+            return self._INTEGER_TYPE
+        elif types == {self._NUMBER_TYPE} or types == {self._INTEGER_TYPE, self._NUMBER_TYPE}:
+            return self._NUMBER_TYPE
+        # to keep backward compatibility with PyArrow, we will not parse types
+        return self._STRING_TYPE
+
+    def _infer_type(self, value: str) -> str:
+        if self._is_boolean(value):
+            return self._BOOLEAN_TYPE
+        elif self._is_integer(value):
+            return self._INTEGER_TYPE
+        elif self._is_number(value):
+            return self._NUMBER_TYPE
+        elif self._is_array(value):
```

Comment:
The check to `_is_array` and `_is_object` can be removed as they have impact given [this](https://github.com/airbytehq/airbyte/pull/29099/files#diff-2ac011610fd803eed6b2f7f1369e9c9ecc6d99b01ebfb475781eec212d081064R263)

Reply:
We agreed on "infer_legacy" so both will be supported in case of `infer_datatypes`

Reply:
This has been updated to have verbose enums describing the different types of inference